File: array/sliding_window/0121_best_time_to_sell_stock.py
# You are given an array prices where prices[i] is the price of a given stock
#  on the ith day.

# You want to maximize your profit by choosing a single day to buy one stock 
# and choosing a different day in the future to sell that stock.

# Return the maximum profit you can achieve from this transaction. If you 
# cannot achieve any profit, return 0.

# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# if prices[i + 1] <= price[i], never buy prices[i]
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        cur_min, max_profit = float('inf'), 0
        for price in prices:
            if price <= cur_min:
                cur_min = price
            else:
                max_profit = max(max_profit, price - cur_min)
        return max_profit

File: array/sliding_window/variable_size/0121_best_time_to_sell_stock.js
/*You are given an array prices where prices[i] is the price of a given stock
 on the ith day.

You want to maximize your profit by choosing a single day to buy one stock 
and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you 
cannot achieve any profit, return 0.
*/
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

// sliding window
// time complexity O(n)
// space complexity O(1)
var maxProfit = function (prices) {
  let left = 0;
  let right = 1;
  let result = 0;
  while (right < prices.length) {
    if (prices[left] < prices[right]) {
      result = Math.max(result, prices[right] - prices[left]);
    } else {
      left = right;
    }
    right++;
  }
  return result;
};

// Array higher order function
// time complexity O(n^2)
// space complexity O(1)
var maxProfit = function (prices) {
  let result = 0;
  while (prices.length > 1) {
    let min = Math.min(...prices);
    let max = Math.max(...prices.slice(prices.indexOf(min)));
    result = Math.max(result, max - min);
    prices.splice(prices.indexOf(min));
  }
  return result;
};

// brute force
// time complexity O(n^2) -- time limit exceeded error
// space complexity O(1)
var maxProfit = function (prices) {
  let result = 0;
  for (let i = 0; i < prices.length - 1; i++) {
    for (let j = i + 1; j < prices.length; j++) {
      const currentProfit = prices[j] - prices[i];
      result = Math.max(max, currentProfit);
    }
  }
  return result;
};
